jwt.js

```diff
@@ -1,61 +1,114 @@
-'use strict'
-
 const fp = require('fastify-plugin')
 const JWT = require('jsonwebtoken')
-const assert = require('assert')
+const steed = require('steed')()
+
+function wrapStaticSecretInCallback (secret) {
+  return function (_, __, cb) {
+    return cb(null, secret)
+  }
+}
 
-function fastifyJWT (fastify, opts, next) {
-  if (!opts.secret) {
+function fastifyJwt (fastify, options, next) {
+  if (!options.secret) {
     return next(new Error('missing secret'))
   }
 
-  const secret = opts.secret
+  let secretCallback = options.secret
+
+  if (typeof secretCallback !== 'function') { secretCallback = wrapStaticSecretInCallback(secretCallback) }
+
+  let _requestProperty = options.userProperty || options.requestProperty || 'user'
+  let _resultProperty = options.resultProperty
+  const credentialsRequired = typeof options.credentialsRequired === 'undefined' ? true : options.credentialsRequired
 
   fastify.decorate('jwt', {
     sign: sign,
     verify: verify,
-    decode: decode,
-    secret: secret
+    secretCallback: secretCallback
   })
 
   next()
 
-  function sign (payload, options, callback) {
-    assert(payload, 'missing payload')
-    options = options || {}
-    if (typeof options === 'function') {
-      callback = options
-      options = {}
-    }
+  function sign (request, reply, next) {
+    let {secret, resultProperty, ...rest} = options
+    steed.waterfall([
+      function getSecret (callback) {
+        secretCallback({}, {}, callback)
+      },
+      function sign (secret, callback) {
+        JWT.sign(request.body, secret, rest, function (err, token) {
+          if (err) {
+            callback(err)
+          } else {
+            callback(null, token)
+          }
+        })
```

Comment:
why is this callback needed? can't you just do `JWT.sign(request.body, secret, rest, callback)`.